sockeye/inference.py

```diff
@@ -1419,32 +1505,49 @@ def _beam_search(self,
         attentions = mx.nd.take(attentions, best_hyp_indices)
         scores_accumulated[:] = mx.nd.take(scores_accumulated, best_hyp_indices)
         finished = mx.nd.take(finished, best_hyp_indices)
+        constraints = [constraints[int(x.asscalar())] for x in best_hyp_indices]
 
-        return sequences, attentions, scores_accumulated, lengths, beam_histories
+        return sequences, attentions, scores_accumulated, lengths, constraints, beam_histories
 
-    def _get_best_from_beam(self,
+    def _get_item_from_beam(self,
                             sequences: mx.nd.NDArray,
                             attention_lists: mx.nd.NDArray,
                             accumulated_scores: mx.nd.NDArray,
                             lengths: mx.nd.NDArray,
-                            beam_histories: Optional[List[BeamHistory]]) -> List[Translation]:
+                            constraints: List[ConstrainedHypothesis],
+                            beam_histories: Optional[List[BeamHistory]],
+                            desired_index: int = 0) -> List[Translation]:
         """
         Return the best (aka top) entry from the n-best list.
 
         :param sequences: Array of word ids. Shape: (batch_size * beam_size, bucket_key).
         :param attention_lists: Array of attentions over source words.
                                 Shape: (batch_size * self.beam_size, max_output_length, encoded_source_length).
         :param accumulated_scores: Array of length-normalized negative log-probs.
+        :param desired_index: which item you want (default: 0, the best)
         :return: Top sequence, top attention matrix, top accumulated score (length-normalized
                  negative log-probs) and length.
         """
         utils.check_condition(sequences.shape[0] == attention_lists.shape[0] \
                               == accumulated_scores.shape[0] == lengths.shape[0], "Shape mismatch")
         # sequences & accumulated scores are in latest 'k-best order', thus 0th element is best
-        best = 0
         result = []
         for sent in range(self.batch_size):
-            idx = sent * self.beam_size + best
+            idx = sent * self.beam_size + desired_index
+
+            # If lexical output constraints were used, there may be some incomplete hypotheses.
+            # This code skips over them.
+            if constraints is not None and constraints[idx] != []:
```

Comment:
This could be shortened to `if constraints is not None and constraints[idx]:`

Reply:
done